src/python/pants/core/util_rules/external_tool.py

```diff
@@ -73,6 +97,15 @@ def my_rule(my_external_tool: MyExternalTool) -> Foo:
     default_version: str
     default_known_versions: List[str]
 
+    # our hard constraints required for this tool, to support the API we expect from it
+    version_constraints: Optional[str] = None
+    constraints: Optional[Requirement] = None
```

Comment:
I'm not understanding how these are different?

Reply:
`constraints` is the parsed version of `version_constraints`. I'm not super happy about it, but I didn't want to parse it every time I referred to it, which it would do if I put it behind a property, for instance. (unless we have a caching property decorator..? )

Reply:
which you see here: https://github.com/pantsbuild/pants/pull/12332#discussion_r670756422
I guess your comment there was wrgt the checking, so not what was in __init__, yet.

Reply:
Ok, I'll be able to get rid of it.